runtime/src/main/java/org/corfudb/runtime/view/LayoutBuilder.java

```diff
@@ -328,37 +330,50 @@ public LayoutBuilder mergePreviousSegment(int segmentIndex) {
+     *                                     moved to the top of sequencer server list due
+     *                                     to being unresponsive or a failed endpoint.
      */
-    public LayoutBuilder assignResponsiveSequencerAsPrimary(@NonNull Set<String> endpoints) {
-
-        List<String> modifiedSequencerServers = new ArrayList<>(layout.getSequencers());
-        for (int i = 0; i < modifiedSequencerServers.size(); i++) {
-            String sequencerServer = modifiedSequencerServers.get(i);
-
-            // Add a sequencer server given sequencerServer is not already included in
-            // endpoints AND it is not included in the unresponsive nodes of the layout.
-            if (!endpoints.contains(sequencerServer) &&
-                !layout.getUnresponsiveServers().contains(sequencerServer)) {
-                modifiedSequencerServers.remove(sequencerServer);
-                modifiedSequencerServers.add(0, sequencerServer);
-                layout.setSequencers(modifiedSequencerServers);
-                return this;
+    public LayoutBuilder assignResponsiveSequencerAsPrimary(@NonNull Set<String> excludedNodes) {
+        List<String> modifiedSequencers = new LinkedList<>(layout.getSequencers());
+        String primarySequencer = null;
+
+        // Consider only tail nodes in the last segment.
+        Set<String> logTails = layout.getLastSegment().getStripes().stream()
+                .map(LayoutStripe::getTailEndpoint)
+                .collect(Collectors.toSet());
+
+        // Elect a sequencer server as primary given that server is not already included
+        // in excludedNodes AND it is not included in the unresponsive nodes of the layout.
+        // Non log tail servers have high priority to be elected for better load distribution.
+        for (String sequencer : modifiedSequencers) {
+            if (excludedNodes.contains(sequencer) ||
+                    layout.getUnresponsiveServers().contains(sequencer)) {
+                continue;
+            }
+            primarySequencer = sequencer;
+            if (!logTails.contains(sequencer)) {
+                break;
             }
         }
 
-        log.warn("Failover to a responsive sequencer server failed. Each of the layout " +
-                 "sequencers:{} is either a failed endpoint or an unresponsive server",
-                 modifiedSequencerServers.toString());
-        throw new LayoutModificationException("All sequencers failed.");
+        if (primarySequencer == null) {
+            log.warn("Failover to a responsive sequencer server failed. Each of the layout " +
+                            "sequencers:{} is either a failed endpoint or an unresponsive server",
+                    modifiedSequencers.toString());
+            throw new LayoutModificationException("All sequencers failed.");
+        }
+
+        modifiedSequencers.remove(primarySequencer);
+        modifiedSequencers.add(0, primarySequencer);
```

Comment:
it can be improved:

```
LinkedList<String> modifiedSequencers = new LinkedList<>(layout.getSequencers());
...
modifiedSequencers.addFirst(primarySequencer);
```

Reply:
Done.